### PR TITLE
Start work on react-guidelines

### PR DESCRIPTION
**Pull request by arhtudormorar**
#### Description of the pull request (what is new / what has changed)
Update documentation for react code guidelines

#### Did you test the changes locally ?
- [x] yes
- [ ] no

#### Which category (categories) does this pull request belong to?
- [ ] document new feature
- [ ] update documentation that is not relevant anymore
- [ ] add examples or more information about a component
- [ ] fix grammar issues
- [x] other

### CODE REVIEW

**Review by razvantomegea — DISMISSED** [review]

File: docs/developers/react-development-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+
+[comment]: # (mx-abstract)
+
+## Introduction
+
+Every developer has his/her own code style that has been developed along the way, with bits and quirks that, at some point, become a part of oneself. 
+
+However, in a big team and in a big project, small quirks and personal preferences can add up and turn the codebase into a big lasagna. 
+
+Given this, we have established some basic principles and a code style we would like to follow in this project. These are, of course, not set in stone, and can be changed, given a valid reason.
```
Comment (CiprianDraghici) [inline docs/developers/react-development-guidelines.md:14]: "to follow in this project" ... which project ? Maybe this can be reformulated to be more generic

File: docs/developers/react-development-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+    props.onClickClick(); 
+} 
+
+<div onClick={handleClick}/> 
+
+//destructured before, instantly known to be from parent 
+<div onClick={onClick}/>`
+```
+
+* We use a system for **branch naming**: \[feature || fix || redesign\]/-\[2-3 words describing the branch\] e.g. `feature/fix-Thing-called-twice`
```
Comment (CiprianDraghici) [inline docs/developers/react-development-guidelines.md:33]: Thing -> thing (lowercase)

File: docs/developers/react-development-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+if (!user) {
+  //do something
+  if (!refetchAttempted){ 
+    refetch();
+  } 
+}
+
+// ✅ DO 
+const user = userSelector(state); 
+if (user == null) {
```
Comment (CiprianDraghici) [inline docs/developers/react-development-guidelines.md:107]: I would use an early return approach here

```
if(user) return;

if(!refetchAttempt) {...}
```

File: docs/developers/react-development-guidelines.md [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+  if (!refetchAttempted) { 
+    refetch();
+  }
+} 
+```
+
+* For folder and file naming we're using the following convention:  
+**camelCase for all folders and files, except when it's a React Component or a Module Root Folder**, in which case we're using PascalCase.  
+Also, for components' and containers' subcomponents, we create separate folders, even if there is no style file present.  
+Each folder that has an exportable component will have an **`index.tsx`** file for ease of import. <br/>
```
Comment (CiprianDraghici) [inline docs/developers/react-development-guidelines.md:118]: no need to be `.tsx`, it's enough to be `.ts`

File: docs/developers/react-development-guidelines.md [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+
+```jsx
+// 🚫 DON'T 
+const handlePress = useCallback(() => setPressed(true)); 
+<TouchableOpacity onPress={handlePress}/> 
+// 🚫 DON'T 
+const value = useMemo(() => user.level * multiplicator); 
+
+// ✅ DO 
+const handlePress = useCallback(() => setPressed(true));
```
Comment (CiprianDraghici) [inline docs/developers/react-development-guidelines.md:190]: missing deps array for useCallback

File: docs/developers/react-development-guidelines.md [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+<Card isFullscreen={true}/> 
+// ✅ DO 
+<Card isFullscreen /> 
+```
+
+* No **`useCallback` or `useMemo` or `React.memo` unless really necessary**. Since the release of hooks, over-optimization has become a big problem.
+
+```jsx
+// 🚫 DON'T 
+const handlePress = useCallback(() => setPressed(true));
```
Comment (CiprianDraghici) [inline docs/developers/react-development-guidelines.md:184]: missing deps array for useCallback

File: docs/developers/react-development-guidelines.md [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+
+1. **Fake modularization**:
+  * Custom hooks may give the impression of modularization, but their logic runs inline in the parent component.
+  * State changes and reactive behavior in the hook will cause a rerender of the host component and any other hooks that depend on the updated hook.
+2. **Hook return values**:
+  * Custom hooks should return either a single function (for a lazy hook) or an object containing a function and some properties;
+  * Avoid returning objects with multiple functions to ensure consistency and maintainability;
+  * Use interfaces for hook return type:
+    ```jsx
+    useMyHook({firstParam, secondParam}: UseMyHookParamsType): UseMyHookReturnType => {}
```
Comment (CiprianDraghici) [inline docs/developers/react-development-guidelines.md:206]: the return type is inferred, I don't think is necessary to set it explicitly

File: docs/developers/react-development-guidelines.md [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+  * Lifting state up should be used sparingly; hooks should primarily be used to gather state in one container and distribute it to child components.
+  * Strive to couple data fetching with UI rendering as isolating as possible to prevent unnecessary rerenders.
+  * Ensure one hook does not trigger the rerender of another by carefully managing dependencies and side effects.
+  * If only 10% of the lines of code in a hook do specific React logic like state management, or calling a selector from Redux, consider splitting the hook into:
+    - a smaller hook that does the React logic and returns the state, and
+    - a statless function that will be called inside the hook and will do the rest of the logic. This stateless function should be tested separately.
+4. **Hook interfaces**:
+  * Use interfaces for hook params if you're passing more than one argument to the hook invocation:
+    ```jsx
+    useMyHook({firstParam, secondParam}: UseMyHookParamsType) => {}
```
Comment (CiprianDraghici) [inline docs/developers/react-development-guidelines.md:218]: Sometimes is simpler to see the props type inline using inline declaration. But it becomes ugly just in case you have a huge object. I assume for 3-4 fields is not necessary to extract a type

**Review by razvantomegea — DISMISSED** [review]

**Review by StanislavSava — DISMISSED** [review]

**Review by CiprianDraghici — DISMISSED** [review]

File: docs/sdk-and-tools/sdk-dapp/internal-processes/guardians.md [hunk 9]
```diff
@@ -1,6 +1,6 @@
 ---
 id: guardians
-title: Guardians
+title: Guardians1
```
Comment (bogdan-rosianu) [inline docs/sdk-and-tools/sdk-dapp/internal-processes/guardians.md:3]: revert this change

File: docs/developers/react-development-guidelines.md [hunk 10]
```diff
@@ -0,0 +1,402 @@
+---
+id: react-development-guidelines
+title: React Development Guidelines
+---
+
+[comment]: # (mx-abstract)
+
+## Introduction
```
Comment (bogdan-rosianu) [inline docs/developers/react-development-guidelines.md:8]: please keep in mind that Docusaurus is different than MD files for handling headings. In the current format, `Introduction` is heading 1, and `Abstracting the logic away into hooks and function` will be a sub-chapter of the `Introduction` in the navigation panel on the right side. 
Screenshot 2023-12-05 at 14 21 09

So use `##` for heading 1, `###` for heading 2 and so on

File: sidebars.js [hunk 11]
```diff
@@ -172,6 +172,7 @@ const sidebars = {
           },
         ],
       },
+      "developers/react-development-guidelines",
```
Comment (bogdan-rosianu) [inline sidebars.js:175]: I don't like the current placing of this page, since I don't consider this page a first level entry in the Developers section:
Screenshot 2023-12-05 at 14 23 51

Instead, I would create a sub-category called `Guidelines` that will refer to this page, as shown in this example: 
Screenshot 2023-12-05 at 14 26 29


In order to do so, replace:
```
"developers/react-development-guidelines",
```

with

```
{
        type: "category",
        label: "Guidelines",
        items: [
          "developers/react-development-guidelines"
        ]
      }
```

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by MiroMargineanu — APPROVED** [review]

**Review by razvantomegea — APPROVED** [review]
